Source IPv6 node IPs as AAAA endpoints

danopia opened the pull request:
**Description**

In an IPv6-enabled cluster, it's possible (and legitiment) for a node's `.status.addresses` list to contain IPv6 addresses. Usually alongside an IPv4 address for dual-stack clusters. Here's a partial example: https://kubernetes.io/docs/tasks/network/validate-dual-stack/#validate-node-addressing

Multiple external-dns sources assume that every IP address is IPv4 and thus has `A` hardcoded as the record type. This results in invalid DNS endpoints being generated. With the node source, this means the sync loop is broken while such a node is still present.

This PR replaces some of the `node.go` logic to support creating either an A endpoint, or an AAAA endpoint, or both at the same time, for each node. Some of the logic might be questionable and I welcome the review.

Fixes #1875 but **does not** actually result in AAAA records being created at this time because external-dns itself does not support AAAA records yet. The planner needs some work to support dual-stack records properly and I'm considering it out of scope of upgrading each source to emit proper endpoints.

So, this PR fixes invalid A endpoints being generated, and replaces them with correct AAAA endpoints that simply aren't actioned on at this point in time.

**Checklist**

- [x] Unit tests updated
- [ ] End user documentation updated

hrvolapeter commented:
It's there any update on the progress? Seems like there weren't any comments from maintainers. This PR seems to mitigate problem with IPv6. Currently external-dns creates A records with IPv6 which fails on cloud provider validation

seanmalloy commented:
/kind feature

danopia commented:
Rebased; there was a conflict from this new line https://github.com/kubernetes-sigs/external-dns/blame/979c2d611bb7a626dfaaade5fe36f7693949a733/source/node.go#L154 which I moved down into the endpoints loop since this changeset can introduce up to two endpoint records per node.

unixfox commented:
Bump

bootc commented:
/remove-lifecycle stale

danopia commented:
I'll finally accept the staleness of this; as of 2022 I have a pretty low appetite for rebasing on this codebase. Hopefully improved AAAA support will eventually merge from other contributions.
